File: answers/other/answers/lintcode/Search in Rotated Sorted Array II.cpp
class Solution {
    /** 
     * param A : an integer ratated sorted array and duplicates are allowed
     * param target :  an integer to be search
     * return : a boolean 
     */
public:
    bool search(vector<int> &A, int target) {
        // write your code here
		size_t left(0);
		size_t right(A.size() - 1);
		while (left <= right && right < A.size()) {
			size_t middle = (left + right) / 2;
			if (A[middle] == target) {
				return true;
			}
			if (A[middle] > A[left]) {
				if (A[left] <= target && target < A[middle]) {
					right = middle - 1;
				}
				else {
					left = middle + 1;
				}
			}
			else if (A[middle] < A[left]) {
				if (A[middle] < target && target <= A[right]) {
					left = middle + 1;
				}
				else {
					right = middle - 1;
				}
			}
			else {
				++ left;
			}
		}
		return false;
    }
};

File: answers/leetcode/Search in Rotated Sorted Array II/Search in Rotated Sorted Array II.java
class Solution {
    public boolean search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            int middle = (left + right) / 2;
            if (nums[middle] == target) {
                return true;
            }
            if (nums[middle] < nums[left] || nums[middle] < nums[right]) {
                if (target > nums[middle] && target <= nums[right]) {
                    left = middle + 1;
                } else {
                    right = middle - 1;
                }
            } else if (nums[middle] > nums[left] || nums[middle] > nums[right]) {
                if (target >= nums[left] && target < nums[middle]) {
                    right = middle - 1;
                } else {
                    left = middle + 1;
                }
            } else {
                left++;
            }
        }
        return false;
    }
}